### PR TITLE
[GlobalIsel] Combine logic of floating point compares

### PR DESCRIPTION
**Pull request by tschuett**
It is purely based on symmetry. Registers can be scalars, vectors, and non-constants.

X < 5.0 || X > 5.0
  ->
X != 5.0

X < Y && X > Y
  ->
  FCMP_FALSE

X < Y && X < Y
  ->
  FCMP_TRUE

see InstCombinerImpl::foldLogicOfFCmps

### CODE REVIEW

**Comment by aemerson** [comment]
Is there a motivating real use case for this? My concern is if we're implementing IR InstCombines in the combiner purely for symmetry we add compile time cost for little benefit.

**Comment by tschuett** [comment]
The motivation was to fold more compares. The DAG combiner has the same combine in `foldLogicOfSetCCs`, but it is tight to the ISD namespace. The Instcombine method of calculating the new predicate is more portable.

**Comment by tschuett** [comment]
I am not interested in feature parity with Instcombine. I am interested in making this combiner more aggressive (up to a point).

Instcombine has for some Opcodes harsher combines than the Dag that are 10 lines of code. I am interested in these 10 lines combines.

**Comment by nikic** [comment]
> Is there a motivating real use case for this? My concern is if we're implementing IR InstCombines in the combiner purely for symmetry we add compile time cost for little benefit.

For DAGCombine, the general policy is that it should only contain combines that are either target-dependent, or arise during legalization/lowering. Everything else should be handled (only) in InstCombine instead. I'd expect GlobalISel to follow the same approach...

File: llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp [hunk 1]
```diff
@@ -6814,12 +6815,90 @@ bool CombinerHelper::tryFoldAndOrOrICmpsUsingRanges(GLogicalBinOp *Logic,
...
+  GFCmp *Cmp2 = getOpcodeDef<GFCmp>(RHS, MRI);
+  if (!Cmp2)
+    return false;
+
+  LLT CmpTy = MRI.getType(Cmp1->getReg(0));
+  LLT CmpOperandTy = MRI.getType(Cmp1->getLHSReg());
+
+  // We build one fcmp, want to fold the fcmps, replace the logic op,
+  // and the fcmps must have the same shape.
+  if (!isLegalOrBeforeLegalizer(
```
Comment (tschuett) [inline llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp:6841]: The result of  `isLegalOrBeforeLaglizer` is target-dependent. Some targets support fcmp with the given types some do not, e.g., `CmpOperandTy` could be a scalable vector.

**Comment by tschuett** [comment]
I always thought DagCombine combines as much as possible, but checks whether the operations are legal on the current device.

**Comment by tschuett** [comment]
InstCombine combines the overflow intrinsics to the fullest. GlobalIsel and the Dag have named opcodes for overflow intrinsics.
GlobalIsel: G_SADDO reads as signed addition overflow.

What is the policy for the GlobalIsel combiner:
* refuse to combine them
* combine the to the fullest with even new tricks
* something in the middle
?

**Review by aemerson — COMMENTED** [review]
If this is trying to mirror DAGCombiner then I think it's ok, it'll likely be useful for GlobalISel too, although I generally prefer seeing some real IR showing benefits.

File: llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp [hunk 2]
```diff
@@ -6814,12 +6815,90 @@ bool CombinerHelper::tryFoldAndOrOrICmpsUsingRanges(GLogicalBinOp *Logic,
   return true;
 }
 
+bool CombinerHelper::tryFoldLogicOfFCmps(GLogicalBinOp *Logic,
+                                         BuildFnTy &MatchInfo) {
+  assert(Logic->getOpcode() != TargetOpcode::G_XOR && "unexpecte xor");
```
[the remark this answers is not in the document]
Comment (tschuett) [reply llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp:6820]: `!=` ! The parameter is `GLogicalBinOp` which includes xor.

File: llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp [hunk 3]
```diff
@@ -6814,12 +6815,90 @@ bool CombinerHelper::tryFoldAndOrOrICmpsUsingRanges(GLogicalBinOp *Logic,
...
+    PredR = CmpInst::getSwappedPredicate(PredR);
+    std::swap(RHS0, RHS1);
+  }
+
+  if (LHS0 == RHS0 && LHS1 == RHS1) {
+    // We determine the new predicate.
+    unsigned CmpCodeL = getFCmpCode(PredL);
+    unsigned CmpCodeR = getFCmpCode(PredR);
+    unsigned NewPred = IsAnd ? CmpCodeL & CmpCodeR : CmpCodeL | CmpCodeR;
+    unsigned Flags = Cmp1->getFlags() | Cmp2->getFlags();
```
[the remark this answers is not in the document]
Comment (tschuett) [reply llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp:6867]: The DAG is very lazy about flags.
https://github.com/llvm/llvm-project/blob/1cbe26d302ca7d3573c1bf43114dfd7d2f0ca476/llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp#L1366

File: llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp [hunk 4]
```diff
@@ -6814,12 +6815,90 @@ bool CombinerHelper::tryFoldAndOrOrICmpsUsingRanges(GLogicalBinOp *Logic,
...
+    // Swap RHS operands to match LHS.
+    PredR = CmpInst::getSwappedPredicate(PredR);
+    std::swap(RHS0, RHS1);
+  }
+
+  if (LHS0 == RHS0 && LHS1 == RHS1) {
+    // We determine the new predicate.
+    unsigned CmpCodeL = getFCmpCode(PredL);
+    unsigned CmpCodeR = getFCmpCode(PredR);
+    unsigned NewPred = IsAnd ? CmpCodeL & CmpCodeR : CmpCodeL | CmpCodeR;
```
[the remark this answers is not in the document]
Comment (tschuett) [reply llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp:6866]: The DAG uses:
```
ISD::CondCode NewCC = IsAnd ? ISD::getSetCCAndOperation(CC0, CC1, OpVT)
                                : ISD::getSetCCOrOperation(CC0, CC1, OpVT);
```

https://github.com/llvm/llvm-project/blob/1cbe26d302ca7d3573c1bf43114dfd7d2f0ca476/llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp#L1363

**Review by aemerson — APPROVED** [review]

**Comment by tschuett** [comment]
Thanks!

**Pull request merged**